### PR TITLE
Ndev 2478 add send erc20 scenario

### PR DESCRIPTION
**Pull request by romanova-natasha**
add send erc20 scenario
add metrics for erc20 scenario
update readme for k6 load tests, minor refactoring

### CODE REVIEW

File: clickfile.py [hunk 1]
```diff
@@ -1265,18 +1265,45 @@ def build(tag):
...
+    k6_prepare_accounts(account_manager, users, balance)
+
+
+@k6.command("deploy-erc20", help="Deploy erc20 contract.")
+@click.option("-n", "--network",required=True, default="local", help="Which network to use for envs assignment")
+@click.option("-b", "--balance", default=None, required=True, help="Initial balance of erc20 owner")
+@click.option("-a", "--bank_account", default=None, required=False, help="Bank account address if needed")
 @catch_traceback
-def run(network, script, users, balance, bank_account):
+def deploy_erc20(network, balance, bank_account):
```
Comment (kristinaNikolaevaa) [inline clickfile.py]: are you sure that we need all this commands to prepare data? 
I suggest to have one command for "run load tests" which will include all needed preparation  (contracts deploy, account creating e.t.c)

File: clickfile.py [hunk 2]
```diff
@@ -1265,18 +1265,45 @@ def build(tag):
...
     
-    print("Running load test scenario...")
+    owner = account_manager.create_account(balance=int(balance))
+    erc20 = deploy_erc20_contract(web3_client, faucet, owner)
+    print(f"ERC20 contract deployed at {erc20.contract.address} with owner {erc20.owner.address}")
+
+
+@k6.command("run", help="Run k6 performance test.")
+@click.option("-n", "--network",required=True, default="local", help="Which network to use for envs assignment")
+@click.option("-s", "--script", required=True, default="./loadtesting/k6/tests/sendNeon.test.js", help="Path to k6 script")
```
Comment (kristinaNikolaevaa) [inline clickfile.py]: Lets make default value = to run all possible load tests for proxy ?

File: loadtesting/k6/readme.md [hunk 3]
```diff
@@ -7,6 +7,11 @@ Make sure you have Golang bin path in PATH, otherwise
 export PATH=$PATH:$(go env GOPATH)/bin
 ```
 
+### Compile contracts
+Make sure you have solc installed and run to generate an abi file for erc20 contract:
```
Comment (kristinaNikolaevaa) [inline loadtesting/k6/readme.md]: can we do it inside the script?

File: loadtesting/k6/readme.md [hunk 4]
```diff
@@ -28,11 +33,27 @@ Install xk6 and build an executable file, tag is a neonlabsorg forked xk6-ethere
...
+
+To run Send ERC20 scenario we need to deploy the contract first and then run load script:
 ```bash
+./clickfile.py k6 deploy-erc20 --network local --balance 200 
+
 ./clickfile.py k6 run --network local --script ./loadtesting/k6/tests/sendNeon.test.js --users 100 --balance 200
 ```
 
-It is mandatory either set up `K6_USERS_NUMBER` and `K6_INITIAL_BALANCE` envs or pass `--users` and `--balance` flags to run the command. We need these values to prepare test accounts for further transfers transactions. 
+It is mandatory either set up `K6_USERS_NUMBER` and `K6_INITIAL_BALANCE` envs or pass `--users` and `--balance` flags to run the command. We need these values to set up options for k6 load scenario.
```
Comment (kristinaNikolaevaa) [inline loadtesting/k6/readme.md]: As I understand you removed the possibility to set K6_USERS_NUMBER and K6_INITIAL_BALANCE
Only option is to pass `--users` and `--balance`

Comment (kristinaNikolaevaa) [inline loadtesting/k6/readme.md]: How can I understand how big the balance should be? 
Maybe it's better to count the value inside and remove the flag ?

File: utils/k6_helpers.py [hunk 5]
```diff
@@ -28,10 +25,30 @@ def k6_prepare_accounts(network, network_object, users, balance, bank_account):
     with open('./loadtesting/k6/data/accounts.json', 'w', encoding='utf-8') as f:
         json.dump(accounts, f)
 
+def deploy_erc20_contract(web3_client, faucet, account):
+    erc20 = ERC20(
```
Comment (kristinaNikolaevaa) [inline utils/k6_helpers.py]: wrong ERC20 ...
It is better to check erc20ForSpl

File: loadtesting/k6/tests/sendErc20.test.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+const sendErc20ErrorCounter = new Counter('send_erc20_errors');
+const sendErc20RequestTime = new Trend('send_erc20_request_time', true);
+
+export const options = sendTokenOptions;
+
+const pathToContractData = '../contracts/ERC20/ERC20';
+const abi = JSON.parse(open(pathToContractData + '.abi'));
+
+export default function sendErc20Test() {
+    // We need to have a unique signer account for each VU
```
Comment (kristinaNikolaevaa) [inline loadtesting/k6/tests/sendErc20.test.js]: I think by this way we will get not real scenario
You can do transferring some tokens to all accounts as part of the preparation step 
and then sent tokens from different senders here

File: loadtesting/k6/tests/utils/ethClient.js [hunk 7]
```diff
@@ -12,11 +12,16 @@ export function ethClient(privateKey) {
...
-    const value = amount;
-    return sendTokens(client, from, to, value, gas, gasPrice, nonce)
+    return sendTokens(client, from, to, amount, null, gas, gasPrice, nonce)
 }
 
-export function sendTokens(client, from, to, value, gas, gasPrice, nonce) {
+export function sendErc20ViaTransferFunction(client, erc20, owner, abi, signerAddress, receiverAddress, amount) {
+    const input = erc20.fillInput(abi, "transfer");
+    const nonce = client.getNonce(signerAddress);
+    return sendTokens(client, owner, receiverAddress, amount, input, null, null, nonce)
```
Comment (kristinaNikolaevaa) [inline loadtesting/k6/tests/utils/ethClient.js]: looks strange 
for erc20 "amount" should be part of input field (or data field) 
and this ""value": value," should be 0 all the time

**Review by kristinaNikolaevaa — CHANGES_REQUESTED** [review]

File: loadtesting/k6/tests/utils/accountManager.js [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+export function createAccount() {
+    var accountKey = wallet.generateKey();
+    return {
+        privateKey: accountKey.private_key,
+        address: accountKey.address,
+    };
+}
+
+export function createFundedAccount(client, amount, bankAccount) {
```
Comment (kristinaNikolaevaa) [inline loadtesting/k6/tests/utils/accountManager.js]: do you use this function somewhere?

File: clickfile.py (hunk 1 above)
Comment (romanova-natasha) [reply clickfile.py]: done

File: loadtesting/k6/readme.md (hunk 3 above)
Comment (romanova-natasha) [reply loadtesting/k6/readme.md]: done

File: loadtesting/k6/readme.md (hunk 4 above)
Comment (romanova-natasha) [reply loadtesting/k6/readme.md]: fixed

File: loadtesting/k6/tests/sendErc20.test.js (hunk 6 above)
Comment (romanova-natasha) [reply loadtesting/k6/tests/sendErc20.test.js]: fixed

File: loadtesting/k6/tests/utils/ethClient.js (hunk 7 above)
Comment (romanova-natasha) [reply loadtesting/k6/tests/utils/ethClient.js]: fixed: full changes of input calculations are in xk6 plugin

File: utils/k6_helpers.py (hunk 5 above)
Comment (romanova-natasha) [reply utils/k6_helpers.py]: we will add erc20ForSpl scenario later

File: loadtesting/k6/readme.md (hunk 4 above)
Comment (romanova-natasha) [reply loadtesting/k6/readme.md]: we will add initial balance calculation/additional funding during the scenario run later

File: clickfile.py (hunk 2 above)
Comment (romanova-natasha) [reply clickfile.py]: I've planned to collect all scenarios for proxy in a common one. Then run it taking into account weight of each separate scenario. Do you want to have it in this branch?

Comment (kristinaNikolaevaa) [reply clickfile.py]: ok, let's do it later

File: loadtesting/k6/tests/sendErc20.test.js [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    const erc20 = client.newContract(erc20Address, JSON.stringify(abi), accountSenderPrivateKey)
+
+    const startTime = new Date();
+
+    try {
+        const receipt = sendErc20ViaTransferFunction(
+            client,
+            erc20,
+            accountSenderAddress,
+            erc20Address,
```
Comment (kristinaNikolaevaa) [inline loadtesting/k6/tests/sendErc20.test.js]: does erc20 object contain its address?

File: loadtesting/k6/tests/sendErc20.test.js [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+            JSON.stringify(abi),
```
Comment (kristinaNikolaevaa) [inline loadtesting/k6/tests/sendErc20.test.js:42]: you can define abi inside the fuction because it can't be different for this function

**Review by kristinaNikolaevaa — APPROVED** [review]

File: loadtesting/k6/tests/sendErc20.test.js (hunk 10 above)
Comment (romanova-natasha) [reply loadtesting/k6/tests/sendErc20.test.js:42]: We use open() function to read the .abi file. This function can be used only at the init stage. Unfortunately we can't move it to ethClient.js file and we can't use smth similar cause we do not have nodejs here

File: loadtesting/k6/tests/sendErc20.test.js (hunk 9 above)
Comment (romanova-natasha) [reply loadtesting/k6/tests/sendErc20.test.js]: No, it is a private field. I removed erc20Address from functions args, I agree it is a bit overloaded here

File: loadtesting/k6/tests/utils/accountManager.js (hunk 8 above)
Comment (romanova-natasha) [reply loadtesting/k6/tests/utils/accountManager.js]: it is removed now

**Pull request merged**